File: project/ConvertToRFMixFormat.R
#### INFO ####
# This script makes RFMix Input from Beagle phased files.
# USAGE: R CMD BATCH -chr ConvertToRFMixFormat.R
# LAST UPDATE: 2010-06-09 01:01:21
#############

## READS input options
rm(list=ls())
chr=commandArgs()[7]
chr=substr(chr, 2, nchar(chr))

admix = read.table(paste("/projects/scratch/ms/beecham/beagle_sept2016/Output/phased.ADMIX_CHR", chr, ".bgl.gz.phased.gz",sep=""), header=T, stringsAsFactors=F)
admix = admix[, -c(1:2)]
admixT = t(admix)
numadmix = nrow(admixT)

eur = read.table(paste("/projects/scratch/ms/beecham/beagle_sept2016/RFMIX_REF/Beagle_ID/EUR_CHR", chr, ".bgl.gz",sep=""), header=T, stringsAsFactors=F)
eur = eur[, -c(1:2)]
eurT = t(eur)
numeur = nrow(eurT)

afr = read.table(paste("/projects/scratch/ms/beecham/beagle_sept2016/RFMIX_REF/Beagle_ID/AFR_CHR", chr, ".bgl.gz",sep=""), header=T, stringsAsFactors=F)
afr = afr[, -c(1:2)]
afrT = t(afr)
numafr = nrow(afrT)

asia = read.table(paste("/projects/scratch/ms/beecham/beagle_sept2016/Output/phased.HGDP_CHR", chr, ".bgl.gz.phased.gz",sep=""), header=T, stringsAsFactors=F)
asia = asia[, -c(1:2)]
asiaT = t(asia)
numasia = nrow(asiaT)

all = rbind(admixT, eurT, afrT, asiaT)

binarize = function(snp){
	asFactor = factor(snp)
	allele1 = levels(asFactor)[1]
	allele2 = levels(asFactor)[2]
	result = rep(NA, length(snp))
	result[snp == allele1] = '0'
	result[snp == allele2] = '1'
	return(result)
}

all01 = apply(all, 2, binarize)

# Transpose back
alleles = t(all01)

write.table(alleles, paste("/projects/scratch/ms/beecham/rfmix_sept2016/alleles/alleles_CHR", chr, ".txt", sep=""), quote=F, sep="", col.names=F, row.names=F)

# Classes
classes = c(rep(0,numadmix), rep(1, numeur), rep(2, numafr), rep(3,numasia))
classesM = matrix(classes, nrow=1)
write.table(classesM, paste("/projects/scratch/ms/beecham/rfmix_sept2016/classes/classes_CHR", chr, ".txt", sep=""), quote=F, row.names=F, col.names=F)

File: project/ConvertToRFMixFormat.py
import numpy as np
import pandas as pd

## READ INPUT FILES ##

admix = open("phased.ADMIX_CHR22.bgl.gz.phased", "r")
eur = open("EUR_CHR22.bgl", "r")
afr = open("AFR_CHR22.bgl", "r")
nam = open("phased.HGDP_CHR22.bgl.gz.phased", "r")

## MAKE FUNCTION TO CREATE MATRIX FROM INPUT FILES ##

def createMatrix(data):
	matrix = []
	n = 0
	
	for line in data:
		splitLine = line.split()
		newline = splitLine[2:]
		if not matrix:
			for sampleID in newline:
				matrix.append([sampleID])
		else:
			i = 0
			while i < len(newline):
				matrix[i].append(newline[i])
				i += 1
		n += 1
		#if n > 5:
			#break
	return matrix

## USE MATRIX FUNCTION ON EACH INPUT FILE ##

admixMatrix = createMatrix(admix)
eurMatrix = createMatrix(eur)
afrMatrix = createMatrix(afr)
namMatrix = createMatrix(nam)

#for a in eurMatrix:
	#print a

## CONVERT EACH MATRIX TO A DATA FRAME ##

admixDataFrame = pd.DataFrame(admixMatrix)
eurDataFrame = pd.DataFrame(eurMatrix)
afrDataFrame = pd.DataFrame(afrMatrix)
namDataFrame = pd.DataFrame(namMatrix)

#print admixDataFrame

## COMPUTE LENGTH OF EACH DATA FRAME ##

admixNum = len(admixDataFrame)
#print admixNum
eurNum = len(eurDataFrame)
#print eurNum
afrNum = len(afrDataFrame)
#print afrNum
namNum = len(namDataFrame)
#print namNum

## ROW BIND ALL DATA FRAMES ##

frames = [admixDataFrame,eurDataFrame,afrDataFrame,namDataFrame]
allDataFrame = pd.concat(frames)

#print allDataFrame

## CREATE FUNCTION TO BINARIZE DATA FRAMES ##

def binarize(snp): # define function

	result = pd.factorize(snp)[0]
	return result

## BINARIZE THE COMBINED DATA FRAME ##

all01 = allDataFrame.apply(binarize,axis=0)

#print all01

## WRITE 'ALLELES' FILE FOR RFMIX INPUT ##

all01.to_csv("ALLELES.txt",sep=' ',header=False,index=False)

## WRITE 'CLASSES' FILE FOR RFMIX INPUT ##

admixClass = [0]*admixNum
eurClass = [1]*eurNum
afrClass = [2]*afrNum
namClass = [3]*namNum

allClass = admixClass + eurClass + afrClass + namClass

thefile = open('CLASSES.txt','w')
for item in allClass:
	thefile.write("%s\n" % item)
	
	


